### PR TITLE
Feature: Recursive directory copying for files in template/

### PR DESCRIPTION
**Pull request by ms14981**
**What:**
- The forge can now copy and transform files that are nested in sub-directories inside the template/ directory.
- Some unit tests written for the forge's JS code.
- Small git, eslint and npm package fixes.

**Why:**
- When starting to create a Java (Maven) generator, we realised that if we want to minimise our users' effort to get an initial working example, we need to output our transformed files in Maven's directory structure (src/main/java/<package>/<file>). This PR adds support for this.

**Caveats:**
- No CI/CD setup for these new tests.

### CODE REVIEW

File: src/generate.js [hunk 1]
```diff
@@ -69,15 +69,90 @@ function getFileName(fileName, tagName = "") {
...
   let result = template(schema);
   log.verbose("Writing to output location");
+  createNestedDirectories(schema, file, outputFolder);
   fs.writeFileSync(
     `${outputFolder}/${getFileName(file, schema._tag?.name)}`,
     result
   );
 }
 
+function processTemplateFactory(
```
Comment (ColinEberhardt) [inline src/generate.js:95]: +1 for this refactor - the `generate` function has started to get a bit unwieldy.

**Review by ColinEberhardt — APPROVED** [review]

**Comment by ColinEberhardt** [comment]
Nice work, the refactor and addition of unit tests are great 👍 

Before merging, can you consider a possible change to the tests? I find the code that mocks the folder structure a little hard to read:

~~~
fs.readdirSync.mockImplementation((filePath, { withFileTypes } = {}) => {
      if (filePath.includes("helpers") || filePath.includes("partials")) {
        return [];
      } else if (withFileTypes) {
        return [
          {
            name: `${fileName}.${fileExtension}`,
            isDirectory: () => false,
          },
        ];
      } else {
        return [`${fileName}.${fileExtension}`];
      }
    });
~~~

How easy do you think it would be to create something more 'visual'. For example:

~~~
const readdirMock = mockReadDir({
   helpers: [],
   partials: [],
   [${fileName}.${fileExtension}]: ""
});
~~~

I know that's not quite right - but basically represent the mocked folder structure as an object literal.

Colin E.

**Comment by ColinEberhardt** [comment]
One other small point:

> we need to output our transformed files in Maven's directory structure (src/main/java//)

I presume you've seen the `-o` CLI option that specifies the output folder? regardless, allowing a nested structure within the template folder is a very useful feature.

File: package.json [hunk 2]
```diff
@@ -6,6 +6,7 @@
   "scripts": {
     "prepare": "husky install",
     "test:generators": "openapi-forge test-generators",
+    "test:javascript": "jest",
```
Comment (ColinEberhardt) [inline package.json:9]: perhaps just make this `test`?

**Comment by ms14981** [comment]
> One other small point:
> 
> > we need to output our transformed files in Maven's directory structure (src/main/java//)
> 
> I presume you've seen the `-o` CLI option that specifies the output folder? regardless, allowing a nested structure within the template folder is a very useful feature.

Yes, the `-o` option will definitely cover most cases. However, it won't cover the case where we might want to generate a pom.xml file in the top level directory, and the rest of the files elsewhere. Currently, the forge will crash if there are any folders nested in the template folder.

**Comment by ms14981** [comment]
> Nice work, the refactor and addition of unit tests are great 👍
> 
> Before merging, can you consider a possible change to the tests? I find the code that mocks the folder structure a little hard to read:
> 
> ```
> fs.readdirSync.mockImplementation((filePath, { withFileTypes } = {}) => {
>       if (filePath.includes("helpers") || filePath.includes("partials")) {
>         return [];
>       } else if (withFileTypes) {
>         return [
>           {
>             name: `${fileName}.${fileExtension}`,
>             isDirectory: () => false,
>           },
>         ];
>       } else {
>         return [`${fileName}.${fileExtension}`];
>       }
>     });
> ```
> 
> How easy do you think it would be to create something more 'visual'. For example:
> 
> ```
> const readdirMock = mockReadDir({
>    helpers: [],
>    partials: [],
>    [${fileName}.${fileExtension}]: ""
> });
> ```
> 
> I know that's not quite right - but basically represent the mocked folder structure as an object literal.
> 
> Colin E.

Good shout, I've made a few changes which hopefully will make those mocks easier to read and work with :)

**Comment by gkocak-scottlogic** [comment]
All seems fine to me! I tried the the branch locally as well 👍

**Pull request merged**